### ISSUE TITLE
A minor typo & a weird behavior with Pytorch 1.5/1.6+NVIDIA 1080Ti that hinders reproducibility

### ISSUE THREAD

**Issue by InnovArul** [oskyhn/CNNs-Without-Borders#6]
Hi,

Thanks for the great work!

I have 2 questions.

1) I believe the code in line: https://github.com/oskyhn/CNNs-Without-borders/blob/master/full_conv/Exp2_Red_Green/red_and_green.py#L56 belongs to 'same' convolution and 'circular' padding. If so, It should be `padding_size = 1`. No?

2) I am facing a weird random behaviours when running two Exp2_Red_Green experiments in parallel (on the same GPU - NVIDIA 1080 Ti) and same pytorch version. (I used pytorch 1.5 and 1.6 to test this behavior).

I run two experiments in parallel as follows:

1st: `python red_and_green.py --conv_type sconv`
2nd: `python red_and_green.py --conv_type fconv`

If I run these experiments alone, the performances are reproduced as mentioned in the paper.
But if I run them in parallel, the experiment results are messed up. i.e., full conv gives way too low numbers than reported in paper. I guess, there is something going on with pytorch/GPU.
If you have your setup with you, could you please run it and kindly let me know if you face similar behavior? 

Thanks!

**Comment by InnovArul** [comment]
**More details:**

1. Running experiments alone:

`python red_and_green.py --conv_type fconv`

results for similar position {0: 90.6, 1: 89.4, 2: 89.5, 3: 90.2, 4: 90.1, 5: 88.9, 6: 89.7, 7: 89.4, 8: 89.9, 9: 89.6}
mean: 89.7300 std: 0.4605 for similar
results for dissimilar position {0: 90.6, 1: 89.4, 2: 89.5, 3: 90.2, 4: 90.1, 5: 88.9, 6: 89.7, 7: 89.4, 8: 89.9, 9: 89.6}
mean: 89.7300 std: 0.4605 for dissimilar

`python red_and_green.py --conv_type sconv`

results for similar position {0: 99.9, 1: 99.8, 2: 99.9, 3: 99.8, 4: 99.9, 5: 99.7, 6: 99.8, 7: 99.7, 8: 99.9, 9: 99.9}
mean: 99.8300 std: 0.0781 for similar
results for dissimilar position {0: 8.2, 1: 9.2, 2: 8.0, 3: 8.3, 4: 9.3, 5: 9.0, 6: 8.1, 7: 6.8, 8: 9.4, 9: 7.6}
mean: 8.3900 std: 0.7943 for dissimilar

2. Running experiments in parallel on same GPU:

The results vary dramatically between multiple runs. Results from one of the runs is as follows:

for 'full conv':

results for similar position {0: 89.5, 1: 89.3, 2: 51.7, 3: 51.7, 4: 51.9, 5: 89.3, 6: 89.6, 7: 90.3, 8: 89.9, 9: 89.8}
mean: 78.3000 std: 17.3724 for similar
results for dissimilar position {0: 89.5, 1: 89.3, 2: 51.7, 3: 51.7, 4: 52.1, 5: 89.3, 6: 89.6, 7: 90.3, 8: 89.9, 9: 89.8}
mean: 78.3200 std: 17.3421 for dissimilar

for 'same conv' & zero pad:

results for similar position {0: 99.9, 1: 99.9, 2: 99.9, 3: 99.8, 4: 99.9, 5: 75.4, 6: 73.6, 7: 85.0, 8: 74.7, 9: 62.2}
mean: 87.0300 std: 13.8345 for similar
results for dissimilar position {0: 8.3, 1: 9.0, 2: 7.7, 3: 8.3, 4: 9.5, 5: 76.5, 6: 73.6, 7: 84.3, 8: 74.3, 9: 63.2}
mean: 41.4700 std: 33.2583 for dissimilar

As we can see, while running experiments in parallel, same conv + zero pad performs better than the results reported in paper (on dissimilar testset). Probably, there is some issue with pytorch/GPU. But it hinders the understanding and reproducibility.

`

**Comment by InnovArul** [comment]
It turns out, the 'best' model weights were saved and loaded again for testing.
That messed up the experiments while running in parallel.

**Comment by oskyhn** [comment]
Hi Arul,

Thanks for comments. You can find the answers of your questions:

1. I used pytorch 1.2 for this experiment. In the experiment, the task is Same-Conv with circular padding, so we need to keep the spatial resolution of feature map as same. If I don't use padding_size=2, then the size of the feature map shrinks (you can check the [documentation](https://pytorch.org/docs/1.2.0/_modules/torch/nn/modules/conv.html#Conv2d)). You can also find similar issues [here](https://github.com/pytorch/pytorch/issues/29712) and [here](https://github.com/pytorch/pytorch/issues/20981). Probably, in the newer versions of pytorch, it has been changed.

2. The experiment itself is small, hence I did not design the experiment for multiple parallel GPUs. Thanks for finding the reason of the problem.

**Comment by InnovArul** [comment]
Thanks a lot for the clarification!
I did not know the issue with 'circular'' padding with pytorch 1.2.0. However, it seems to have been fixed in 1.6.0.
About 2nd point, it is understandable. I did not go thru the code before posting the question. sorry for that.

I have one more question:
I notice that you have used 'Conv with stride 2' in red_green classification experiments rather than 'conv+maxpool' blocks. Is there any specific reason for this design choice? 
I am curious if we can expect this observation of 'full' convolution being translation equivariant than other types of convolution (valid and same) types hold in the case of 'conv+maxpool' blocks as well?

Thanks!
